### DISCUSSION TITLE
Average Bug age

### DISCUSSION THREAD

**Discussion by aayaffe** [nbrown02/FlowViz#102] [Ideas]
When measuring a successful team one of the big measures I believe should be used is the amount of time a bug was open. It is a measure of how fast an issue is resolved for the customer. 

It would be great if a specific graph of average open bug age would exist in this dashboard. 

Lead time is quite close to it but it does not take into account the open issues and WIP only deals with items that are currently in progress. Furthermore currently there exist only one graph for all types of WIs

Thank you for a wonderful dashboard.

**Comment by nbrown02** [comment]
Thanks for sharing - I have moved this to the ideas section so as not to be conflated with bugs on the template :)

So, if I understand correctly, this is the date difference between when a bug was created (CreatedDate) and the current date? And this is for everything not in a completed state?

**Reply by aayaffe** [reply]
Was not aware of a discussions page. Thank you.
This is indeed the difference between the created and current date for open bugs, but for closed bugs it is the difference between created and completed.

**Reply by nbrown02** [reply]
Thanks - so it's essentially the AVG of this column in yellow?
![image](https://github.com/nbrown02/FlowViz/assets/29369962/e5af6f47-74ad-4705-b1b7-2ad77cc97236)

**Reply by aayaffe** [reply]
Seems exactly right

**Reply by nbrown02** [reply]
So the issue would be what would be your time element on your x-axis? At the minute, the dates table has a relationship with work items when they are started or completed. By including things that are open (and not started or completed) it becomes tricky visualizing this. 

What you could do, if you just wanted a static 'number' would be:

1) Click 'Transform Data' and find your _WorkItems In Progress2_ table and adjust the 'Source' to included 'CreatedDate'
![image](https://github.com/nbrown02/FlowViz/assets/29369962/1851a51b-0bd6-43d8-8acf-e806a95d2477)

2) Hit 'Close & Apply' (top left)

3) Once loaded, in your _WorkItems In Progress2_ table, add in a column called TimeSinceCreated:
`TimeSinceCreated =
DATEDIFF ( 'WorkItems In Progress2'[CreatedDate], TODAY (), DAY )
`

4) Create a new table called 'TimeOpen' with the following formula
`TimeBugIsOpen = 
UNION (
    SELECTCOLUMNS (
        FILTER('WorkItems Completed', [WorkItemType] = "Bug"),
        "WorkItemId", [WorkItemId],
        "WorkItemType", [WorkItemType],
        "LeadTimeOrOpen", [LeadTimeDays]
    ),
    SELECTCOLUMNS (
        FILTER('WorkItems In Progress2', [WorkItemType] = "Bug"),
        "WorkItemId", [WorkItemId],
        "WorkItemType", [WorkItemType],
        "LeadTimeOrOpen", [TimeSinceCreated]
    )
)`

![image](https://github.com/nbrown02/FlowViz/assets/29369962/57d5915c-0c75-4bc2-9b2a-c99e4cabdc46)

5) Add to a number visual and then create an average of this
![image](https://github.com/nbrown02/FlowViz/assets/29369962/1414877f-e14c-4b6b-a897-a2e405835e03)

**Reply by aayaffe** [reply]
Thank you very much for the comprehensive reply! 🙏
It works very well. It is only missing the element of history.
What I did is add another column which is called DateClosedOrTodaySK which helps with visualizing it on the time line.
Do you think something like that can be implemented in your template?
The main issue will be connecting it to the WorkItems table because of relationship issues with the Date table. I am not sure how can that be solved.

**Reply by nbrown02** [reply]
So I won't be looking to add it into the template, just as it's the first time I've had a request like this and would mean a design change if it were to be a new chart, I have however just thought of how you can do it and solve your issue :)

1) Make a copy of _WorkItems WIP Age_ (let's say this is called  _WorkItems Bug Age_)

2) Change the 'source' to be just WorkItemType eq Bug and include CreatedDate and LeadTimeDays as columns to group by

3) Click close and apply

4) Create a relationship between Dates table and DateSK in WorkItems Bug Age (this solves our date issue)

5) Add a column `DaysOpen = IF('WorkItems WIP Age'[LeadTimeDays] <> BLANK(), 'WorkItems WIP Age'[LeadTimeDays], 'WorkItems WIP Age'[DateValue]-'WorkItems WIP Age'[CreatedDate])`

6) Create a measure `Avg. Days Open = AVERAGE('WorkItems WIP Age'[DaysOpen])`

7) Use 'Dates'[Date] for your x-axis and 'WorkItems Bug Age'[Avg. Days Open] for your y-axis:

![image](https://github.com/nbrown02/FlowViz/assets/29369962/b08aa12f-0fa5-486f-8008-7118ac692380)

**Reply by aayaffe** [reply]
This seems like a very good solution.
I have a little problem. When adding LeadTimeDays to the groupby field in the odata query, it return all null values. Do you have any Idea what can cause that?
Thank you.

**Comment by nbrown02** [comment]
It will be null until the day it is Done. As that table is showing you that
value on a particular day. If I had an item in progress yesterday the
LeadTimeDays would be 0 but if I moved it to done today then that column
will populate (as the item is done). If that makes sense?

**Reply by aayaffe** [reply]
I am not sure.
Does that mean that I should include in the basic query items that are in the done state?

**Comment by nbrown02** [comment]
I'd say yes :)
